core/services/gateway/handlers/vault/handler.go

```diff
@@ -80,152 +99,191 @@ type SecretEntry struct {
+		h.lggr.Errorw("no pending request found for ID", "request_id", resp.ID)
+		h.metrics.requestInternalError.Add(ctx, 1, metric.WithAttributes(
+			attribute.String("don_id", h.donConfig.DonId),
+			attribute.String("error", api.StaleNodeResponseError.String()),
+		))
+		return nil
 	}
-	defer delete(h.pendingRequests, resp.ID)
 
 	rawResponse, err := jsonrpc.EncodeResponse(resp)
 	if err != nil {
-		return h.sendResponse(ctx, h.errorResponse(pendingRequest.req, api.NodeReponseEncodingError, fmt.Errorf("failed to marshal response: %w", err)), pendingRequest.callbackCh)
+		return h.sendResponse(ctx, ur, h.errorResponse(ur.req, api.NodeReponseEncodingError, fmt.Errorf("failed to marshal response: %w", err)))
 	}
-	responseObj := gw_handlers.UserCallbackPayload{
-		RawResponse: rawResponse,
-		ErrorCode:   api.NoError,
+
+	var errorCode api.ErrorCode
+	if resp.Error != nil {
+		errorCode = api.FromJSONRPCErrorCode(resp.Error.Code)
+	} else {
+		errorCode = api.NoError
 	}
 
-	select {
-	case pendingRequest.callbackCh <- responseObj:
-		promRequestSuccess.WithLabelValues(h.donConfig.DonId).Inc()
-		h.lggr.Infof("Processed response for request %s from node %s", resp.ID, nodeAddr)
-		return nil
-	case <-ctx.Done():
-		promRequestInternalError.WithLabelValues(h.donConfig.DonId, api.RequestTimeoutError.String()).Inc()
-		return ctx.Err()
+	successResp := gw_handlers.UserCallbackPayload{
+		RawResponse: rawResponse,
+		ErrorCode:   errorCode,
 	}
+	return h.sendResponse(ctx, ur, successResp)
 }
 
-func (h *handler) handleSecretsCreate(ctx context.Context, jsonRequest *jsonrpc.Request[json.RawMessage], callbackCh chan<- gw_handlers.UserCallbackPayload) error {
-	var req SecretsCreateRequest
-	if err := json.Unmarshal(*jsonRequest.Params, &req); err != nil {
-		return h.sendResponse(ctx, h.errorResponse(jsonRequest, api.UserMessageParseError, err), callbackCh)
+func (h *handler) handleSecretsCreate(ctx context.Context, ar activeRequest) error {
+	var secretsCreateRequest SecretsCreateRequest
+	if err := json.Unmarshal(*ar.req.Params, &secretsCreateRequest); err != nil {
+		return h.sendResponse(ctx, ar, h.errorResponse(ar.req, api.UserMessageParseError, err))
 	}
 
-	if req.ID == "" || req.Value == "" {
-		return h.sendResponse(ctx, h.errorResponse(jsonRequest, api.InvalidParamsError, errors.New("secret id and value cannot be empty")), callbackCh)
+	if secretsCreateRequest.ID == "" || secretsCreateRequest.Value == "" {
+		return h.sendResponse(ctx, ar, h.errorResponse(ar.req, api.InvalidParamsError, errors.New("secret id and value cannot be empty")))
 	}
 
-	h.mu.Lock()
-	h.pendingRequests[jsonRequest.ID] = &pendingRequest{
-		callbackCh: callbackCh,
-		req:        jsonRequest,
-	}
-	h.mu.Unlock()
-
 	// At this point, we know that the request is valid and we can send it to the nodes
 	var nodeErrors []error
 	for _, node := range h.donConfig.Members {
-		err := h.don.SendToNode(ctx, node.Address, jsonRequest)
+		h.mu.RLock()
```

Comment:
Why do we need to acquire the mutex here? The SendToNode call could block for a long time.